Mismatch between missing keys if state dict hooks are used

### System Info

The behavior in `torch.load_model` has changed to error if there is a mismatch between the keys in the loaded state dict and the file, even if strict loading is turned off. It's between 0.5.3 and 0.6, but possibly over a few commits. https://github.com/huggingface/safetensors/commit/c7b18fb2f674a6d35f6a6cec4377061488b9c5c0

https://github.com/huggingface/safetensors/blob/2b253406f8896a5589083f68d05e5fbda30778a6/bindings/python/py_src/safetensors/torch.py#L257-L286

The exception on L271 is not raised if `strict = False`, however pytorch allows users to [define hooks](https://docs.pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.register_load_state_dict_pre_hook) to modify the state dict before it's loaded into the model. For example, we've used this to remove a prefix from all the tensor names (e.g. `model.head` -> `head`).  This hook is called in `model.load_state_dict` (L273), after safetensors has identified what keys are missing.

This creates a mismatch between `torch_missing` (which is empty, because our loaded dict matches the keys) and `missing` (which is populated earlier on). Earlier versions of the code respect `strict`. It also means that the model still has to load before this happens, so we don't get any efficiency gain from checking prior to the call to `load_state_dict` (which seems to be why this change was introduced?)

If this is not permitted then I would flag the presence of any registered model hook as potentially incompatible. You can check if the ordered dict `model._load_state_dict_pre_hooks` is empty or not.

In our case we can re-upload weights with fixed-up names.

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Reproduction

See as an example: https://github.com/weecology/DeepForest/issues/1099

The hook is here: https://github.com/weecology/DeepForest/blob/d0d6c0a7310dad417d0454ec3e36acf81dfe7340/src/deepforest/models/retinanet.py#L98

```python
       torch_missing, torch_unexpected = model.load_state_dict(state_dict, strict=False)
        # Sanity check that the work we've done matches
        # Pytorch internal loading.
        torch_missing = set(torch_missing)
        torch_unexpected = set(torch_unexpected)
        for to_remove_group in to_removes.values():
            for to_remove in to_remove_group:
                if to_remove not in torch_missing:
                    torch_unexpected.add(to_remove)
                else:
                    torch_missing.remove(to_remove)
>       assert torch_missing == missing, f"{torch_missing} != {missing}"
               ^^^^^^^^^^^^^^^^^^^^^^^^
E       AssertionError: set() != {'backbone.body.layer3.1.bn3.bias', 'backbone.body.layer1.1.bn3.weight', 'backbone.body.layer1.2.conv1.weight', 'backbone.body.layer3.0.downsample.1.weight', 'backbone.body.layer3.1.bn2.weight', 'backbone.body.layer1.0.bn1.running_mean', 'backbone.fpn.layer_blocks.0.0.bias', 'head.regression_head.conv.0.0.bias', 'backbone.body.layer3.3.conv2.weight', 'backbone.body.layer1.1.bn2.running_var', 'backbone.body.layer1.2.bn2.bias', 'backbone.body.layer3.5.bn2.running_var', 'backbone.body.bn1.bias', 'backbone.body.layer2.1.bn3.running_mean', 'backbone.body.layer3.5.conv2.weight', 'backbone.body.layer2.0.bn2.running_var', 'backbone.body.layer3.0.bn3.weight', 'backbone.body.layer3.5.bn1.weight', 'backbone.body.layer4.0.bn3.bias', 'backbone.body.layer2.3.bn2.running_var', 'backbone.body.layer1.2.bn1.running_mean', 'backbone.body.layer2.0.bn2.running_mean', 'backbone.body.layer3.3.conv1.weight', 'backbone.body.layer3.5.bn1.bias', 'backbone.body.layer1.0.downsample.0.weight', 'backbone.body.layer3.2.bn1.weight', 'backbone.body.layer3.1.conv2.weight', 'backbone.body.layer3.0.conv1.weight', 'backbone.body.layer3.1.conv3.weight', 'backbone.body.layer4.1.conv2.weight', 'backbone.fpn.inner_blocks.2.0.weight', 'backbone.body.layer2.2.conv2.weight', 'backbone.body.layer3.1.bn2.running_mean'
```

### Expected behavior

Error should not be raised if model.load_state_dict returns no missing/unexpected entries (or if strict is false).

--- a/bindings/python/py_src/safetensors/torch.py
+++ b/bindings/python/py_src/safetensors/torch.py
@@ -221,68 +221,23 @@ def load_model(
     to_removes = _remove_duplicate_names(
         model_state_dict, preferred_names=state_dict.keys()
     )
-
-    reverse_to_remove = {}
-    for key, to_remove_group in to_removes.items():
-        for to_remove in to_remove_group:
-            reverse_to_remove[to_remove] = key
-
-    # We iterate on the model, so we'll add keys we find missing
-    # here
-    missing = set()
-    # We start with all keys on disk declared as unexpected, we'll
-    # slowly remove them when we find them
-    unexpected = set(state_dict.keys())
-    # Some keys can be invalid too.
-    invalid = set()
-
-    for k, mv in model_state_dict.items():
-        actual_k = reverse_to_remove.get(k, None)
-        if actual_k is not None:
-            look_k = actual_k
-        else:
-            look_k = k
-        v = state_dict.get(look_k, None)
-        if v is None:
-            missing.add(k)
-        else:
-            # We can actually check for the shapes while we're at it.
-            # For the device, it's trickier given torch's internals
-            # There might be some Meta device for faster initiation
-            if v.dtype != mv.dtype or v.shape != mv.shape:
-                invalid.add(k)
-            if actual_k is None:
-                unexpected.remove(k)
-
+    missing, unexpected = model.load_state_dict(state_dict, strict=False)
     missing = set(missing)
-    unexpected = set(unexpected)
-    if strict and (missing or unexpected or invalid):
+    for to_remove_group in to_removes.values():
+        for to_remove in to_remove_group:
+            if to_remove not in missing:
+                unexpected.append(to_remove)
+            else:
+                missing.remove(to_remove)
+    if strict and (missing or unexpected):
         missing_keys = ", ".join([f'"{k}"' for k in sorted(missing)])
         unexpected_keys = ", ".join([f'"{k}"' for k in sorted(unexpected)])
-        invalid_keys = ", ".join([f'"{k}"' for k in sorted(invalid)])
         error = f"Error(s) in loading state_dict for {model.__class__.__name__}:"
         if missing:
             error += f"\n    Missing key(s) in state_dict: {missing_keys}"
         if unexpected:
             error += f"\n    Unexpected key(s) in state_dict: {unexpected_keys}"
-        if invalid:
-            error += f"\n    Invalid key(s) in state_dict: {invalid_keys}, mismatched dtypes or shape."
-        del state_dict
         raise RuntimeError(error)
-
-    torch_missing, torch_unexpected = model.load_state_dict(state_dict, strict=False)
-    # Sanity check that the work we've done matches
-    # Pytorch internal loading.
-    torch_missing = set(torch_missing)
-    torch_unexpected = set(torch_unexpected)
-    for to_remove_group in to_removes.values():
-        for to_remove in to_remove_group:
-            if to_remove not in torch_missing:
-                torch_unexpected.add(to_remove)
-            else:
-                torch_missing.remove(to_remove)
-    assert torch_missing == missing, f"{torch_missing} != {missing}"
-    assert torch_unexpected == unexpected, f"{torch_unexpected} != {unexpected}"
     return missing, unexpected
 
 

